File: python/total_spend_test.py
from transaction import Transaction
from invoice_item import InvoiceItem
from total_spend import TotalSpend

def test_by_spend_type1():
  txn_1 = Transaction(name="Insurance premium", category="Insurance", amount=100)
  itm_1 = InvoiceItem(spend_type="reimbursement", amount=100, transaction=txn_1)

  txn_2 = Transaction(name="Dentist appointment", category="copay", amount=11.33)
  itm_2 = InvoiceItem(spend_type="overage", amount=11.33, transaction=txn_2)

  total = TotalSpend.by_spend_type([itm_1, itm_2])
  
  assert total == {"reimbursement": 100, "overage": 11.33}


def test_by_spend_type2():
  txn_1 = Transaction(name="Insurance premium", category="Insurance", amount=50.75)
  itm_1 = InvoiceItem(spend_type="reimbursement", amount=50.75, transaction=txn_1)

  txn_2 = Transaction(name="Dentist appointment", category="copay", amount=91.25)
  itm_2 = InvoiceItem(spend_type="reimbursement", amount=49.25, transaction=txn_2)
  itm_3 = InvoiceItem(spend_type="overage", amount=42, transaction=txn_2)

  total = TotalSpend.by_spend_type([itm_1, itm_2, itm_3])
  
  assert total == {"reimbursement": 100.0, "overage": 42.0}


def test_by_spend_type_and_category():
  txn_1 = Transaction(name="Insurance premium", category="Insurance", amount=50.75)
  itm_1 = InvoiceItem(spend_type="reimbursement", amount=50.75, transaction=txn_1)

  txn_2 = Transaction(name="Dentist appointment", category="copay", amount=91.25)
  itm_2 = InvoiceItem(spend_type="reimbursement", amount=49.25, transaction=txn_2)
  itm_3 = InvoiceItem(spend_type="overage", amount=42, transaction=txn_2)

  txn_3 = Transaction(name="Target purchase", category="retail", amount=5.25)
  itm_4 = InvoiceItem(spend_type="overage", amount=5.25, transaction=txn_3)

  txn_4 = Transaction(name="Amazon purchase", category="online", amount=10.0)
  itm_5 = InvoiceItem(spend_type="overage", amount=10.0, transaction=txn_4)

  total = TotalSpend.by_spend_type_and_category([itm_1, itm_2, itm_3, itm_4, itm_5])

  assert total == {("reimbursement", "Insurance"): 50.75, ("reimbursement", "Copay"): 49.25, ("overage", "Copay"): 42.0, ("overage", "Other"): 15.25}


if __name__ == "__main__":
  test_by_spend_type1()
  test_by_spend_type2()
  test_by_spend_type_and_category()

  print("All tests passed!")

File: dotnet/MiniThatchTest/TotalSpendTest.cs
namespace MiniThatchTest;
using MiniThatch;

public class TotalSpendTest : UnitTest
{
    protected override List<Action> Tests => new()
    {
        TestBySpendType1,
        TestBySpendType2,
        TestBySpendTypeAndCategory,
    };

    private void TestBySpendType1()
    {
        var txn1 = new Transaction("Insurance premium", "Insurance", 100);
        var itm1 = new InvoiceItem("reimbursement", 100, txn1);

        var txn2 = new Transaction("Dentist appointment", "copay", 11.33);
        var itm2 = new InvoiceItem("overage", 11.33, txn2);

        var total = TotalSpend.BySpendType(new List<InvoiceItem> { itm1, itm2 });

        AssertEqual(total, new Dictionary<string, double> { { "reimbursement", 100 }, { "overage", 11.33 } });
    }

    private void TestBySpendType2()
    {
        var txn1 = new Transaction("Insurance premium", "Insurance", 50.75);
        var itm1 = new InvoiceItem("reimbursement", 50.75, txn1);

        var txn2 = new Transaction("Dentist appointment", "copay", 91.25);
        var itm2 = new InvoiceItem("reimbursement", 49.25, txn2);
        var itm3 = new InvoiceItem("overage", 42, txn2);

        var total = TotalSpend.BySpendType(new List<InvoiceItem> { itm1, itm2, itm3 });

        AssertEqual(total, new Dictionary<string, double> { { "reimbursement", 100 }, { "overage", 42 } });
    }

    private void TestBySpendTypeAndCategory()
    {
        var txn1 = new Transaction("Insurance premium", "Insurance", 50.75);
        var itm1 = new InvoiceItem("reimbursement", 50.75, txn1);

        var txn2 = new Transaction("Dentist appointment", "copay", 91.25);
        var itm2 = new InvoiceItem("reimbursement", 49.25, txn2);
        var itm3 = new InvoiceItem("overage", 42, txn2);

        var txn3 = new Transaction("Target purchase", "retail", 5.25);
        var itm4 = new InvoiceItem("overage", 5.25, txn3);

        var txn4 = new Transaction("Amazon purchase", "online", 10.0);
        var itm5 = new InvoiceItem("overage", 10.0, txn4);

        var total = TotalSpend.BySpendTypeAndCategory(new List<InvoiceItem> { itm1, itm2, itm3, itm4, itm5 });

        AssertEqual(total, new Dictionary<string, Dictionary<string, double>>
        {
            {
                "reimbursement", new Dictionary<string, double>
                {
                    { "Insurance", 50.75 },
                    { "Copay", 49.25 },
                }
            },
            {
                "overage", new Dictionary<string, double>
                {
                    { "Copay", 42 },
                    { "Other", 15.25 },
                }
            },
        });
    }
}
